TELCODOCS 477 CNF-3882 4.12 ACM Topology Aware Lifecycle Manager (#50952)
openshift/openshift-docs: modules/cnf-topology-aware-lifecycle-manager-about-cgu-crs.adoc

* Actions to perform before and after the update
* Update timing

You can control the start time of an update using the `enable` field in the `ClusterGroupUpgrade` CR.
For example, if you have a scheduled maintenance window of four hours, you can prepare a `ClusterGroupUpgrade` CR with the `enable` field set to `false`.

You can set the timeout as follows:
[source,yaml]
----
remediationStrategy:
        maxConcurrency: 1
        timeout: 240
----

> You can use the `batchTimeoutAction` to determine what happens if an upgrade fails for a cluster.

Reviewer: update/upgrade

Author: corrected to "update"

> You can specify `continue` to skip the failing cluster and continue to upgrade other clusters, or `abort` to stop the upgrade for all clusters.

Reviewer: The second part of this sentence feels misleading. We don't actually stop anything other the CGU itself, i.e. stop policy remediation on all clusters

Author: ok, I've reworded "....`abort` to stop policy remediation for all clusters."

  uid: cca245a5-4bca-45fa-89c0-aa6af81a596c
Spec:
  actions:
    afterCompletion:
      deleteObjects: true
    beforeEnable: {}
  backup: false
  clusters: <1>
    - spoke1
  enable: true <2>
  managedPolicies: <3>
    - talm-policy
  preCaching: false
  remediationStrategy: <4>
    canaries: <5>
        - spoke1
    maxConcurrency: 2 <6>
    timeout: 240
  clusterLabelSelectors: <7>
    - matchExpressions:
      - key: label1
      operator: In
      values:
        - value1a
        - value1b
  batchTimeoutAction: <8>
status: <9>
    clusters:
      - name: spoke1
        state: complete
      - currentPolicy:
          name: talm-policy
          status: NonCompliant
        name: spoke2
        state: timedout
    computedMaxConcurrency: 2
    conditions:
      - lastTransitionTime: '2022-11-18T16:27:15Z'
        message: All selected clusters are valid
        reason: ClusterSelectionCompleted
        status: 'True'
        type: ClustersSelected <10>
      - lastTransitionTime: '2022-11-18T16:27:15Z'
        message: Completed validation
        reason: ValidationCompleted
        status: 'True'
        type: Validated <11>
      - lastTransitionTime: '2022-11-18T16:37:16Z'

>         message: The cluster backup is in progress
>         reason: NotStarted
>         status: 'False'
>         type: Progressing

Reviewer: Why do we show a timedout example here? Also, The Progressing condition doesn't look right. It shouldn't timeout while we are still doing backup.
```suggestion
      message: Policy remediation took too long
      reason: TimedOut
      status: 'False'
      type: Progressing
```

Author: ok

      - lastTransitionTime: '2022-11-18T16:37:16Z'
        message: Policy remediation took too long
        reason: TimedOut
        status: 'False'
        type: Succeeded
    managedPoliciesForUpgrade:
      - name: talm-policy
        namespace: talm-namespace
    managedPoliciesNs:
      talm-policy: talm-namespace
    remediationPlan:
      - - spoke1

>         - spoke2

Reviewer: if spoke1 is canary, we would have two batches even though maxConcurrency is 2.

Author: done

    status:
      currentBatch: 1
      currentBatchRemediationProgress:
        spoke1:
          state: Completed
        spoke2:
          policyIndex: 0

>           state: InProgress

Reviewer: These lines should be removed if we are showing a succeeded or timeout CGU

Author: ok

      currentBatchStartedAt: '2022-11-18T16:27:16Z'
      startedAt: '2022-11-18T16:27:15Z'

----
<1> Defines the list of clusters to update.
<2> The `enable` field is set to `true`.
<3> Lists the user-defined set of policies to remediate.
<4> Defines the specifics of the cluster updates.
<5> Defines the clusters for canary updates.
<6> Defines the maximum number of concurrent updates in a batch. The number of remediation batches is the number of canary clusters, plus the number of clusters, except the canary clusters, divided by the maxConcurrency value. The clusters that are already compliant with all the managed policies are excluded from the remediation plan.
<7> Displays the parameters for selecting clusters.
<8> Controls what happens if a batch times out. Possible values are `abort` or `continue`. If unspecified, the default is `continue`.
<9> Displays information about the status of the updates.
<10> The status for `clustersSelected` is `true`.

> <11> The status for `Validated` is `true`.

Reviewer: I see better notes for these two in later examples. Move them here?

Author: done

For more information see the "Using the container image pre-cache feature" section.

[id="creating_backup"]
== Creating a backup
For {sno}, {cgu-operator} can create a backup of a deployment before an update. If the update fails, you can recover the previous version and restore a cluster to a working state without requiring a reprovision of applications. To use the backup feature you first create a `ClusterGroupUpgrade` CR with the `backup` field set to `true`. To ensure that the contents of the backup are up to date, the backup is not taken until you set the `enable` field in the `ClusterGroupUpgrade` CR to `true`.

{cgu-operator} uses the `BackupSucceeded` condition to report the status and reasons as follows:

* `true`
+
Backup is completed for all clusters or the backup run has completed but failed for one or more clusters. If backup fails for any cluster, the update fails for that cluster but proceeds for all other clusters.
* `false`
+
Backup is still in progress for one or more clusters or has failed for all clusters.

For more information, see the "Creating a backup of cluster resources before upgrade" section.

[id="updating_clusters"]

> == Updating clusters

Reviewer: An inprogress example for this section

Reviewer: We talked about this, but let me know if you need more help on this example

{cgu-operator} enforces the policies following the remediation plan.
Enforcing the policies for subsequent batches starts immediately after all the clusters of the current batch are compliant with all the managed policies. If the batch times out, {cgu-operator} moves on to the next batch. The timeout value of a batch is the `spec.timeout` field divided by the number of batches in the remediation plan.

{cgu-operator} uses the `Progressing` condition to report the status and reasons as follows:

* `true`
+
{cgu-operator} is remediating non-compliant policies.
* `false`
+
The update is not in progress. Possible reasons for this are:
+
** All clusters are compliant with all the managed policies.
** The update has timed out as policy remediation took too long.
** Blocking CRs are missing from the system or have not yet completed.

> 

Reviewer: ** CGU is not enabled
** Backup is still in progress

Author: ok

      - policy1-common-cluster-version-policy
      - policy2-common-pao-sub-policy
      remediationStrategy:
        maxConcurrency: 1
        timeout: 240
    status:
      clusters:
        - name: spoke1
          state: complete
        - name: spoke4
          state: complete
      conditions:
      - message: All selected clusters are valid
        reason: ClusterSelectionCompleted
        status: "True"
        type: ClustersSelected <1>
      - message: Completed validation
        reason: ValidationCompleted
        status: "True"
        type: Validated <2>
      - message: All clusters are compliant with all the managed policies
        reason: Completed
        status: "False"
        type: Progressing <3>
      - message: All clusters are compliant with all the managed policies
        reason: Completed
        status: "True"
        type: Succeeded <4>
      managedPoliciesForUpgrade:
      - name: policy1-common-cluster-version-policy
        namespace: default
      - name: policy2-common-pao-sub-policy
        namespace: default
      remediationPlan:
      - - spoke1
      - - spoke4
      status:
        currentBatchRemediationProgress:
          spoke4:

>             state: Completed <5>

Reviewer: This part is removed if CGU is done. And there should be start and complete time.

Author: ok

----
<1> The `ClustersSelected` fields show that all selected clusters are valid.

> <2> The `Validated` fields show that all selected clusters have been validated.

Reviewer: Not sure if we need to repeat these two in every example.

Author: ok - removed

<3> In the `Progressing` fields, the status is `false` as the update has completed; clusters are compliant with all the managed policies.
<4> The `Succeeded` fields show that the validations completed successfully.
<5> Displays that all the policies are applied to the cluster.

The status field includes a list of clusters and their respective status. The status of a cluster can be `complete` or `timedout`. If a cluster is `timedout`, the `currentPolicy` field shows the name of the policy that was being applied and the status of that policy.

.Sample `ClusterGroupUpgrade` CR in the `timedout` state

[source,yaml]
----
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  creationTimestamp: '2022-11-18T16:27:15Z'
  finalizers:
    - ran.openshift.io/cleanup-finalizer
  generation: 1
  name: talm-cgu
  namespace: talm-namespace
  resourceVersion: '40451823'
  uid: cca245a5-4bca-45fa-89c0-aa6af81a596c
spec:
  actions:
    afterCompletion:
      deleteObjects: true
    beforeEnable: {}
  backup: false
  clusters:
    - spoke1
    - spoke2
  enable: true
  managedPolicies:
    - talm-policy
  preCaching: false
  remediationStrategy:
    maxConcurrency: 2

>     timeout: 9

Author: ok

      state: complete
    - currentPolicy: <1>
        name: talm-policy
        status: NonCompliant
      name: spoke2
      state: timedout
  computedMaxConcurrency: 2
  conditions:
    - lastTransitionTime: '2022-11-18T16:27:15Z'
      message: All selected clusters are valid
      reason: ClusterSelectionCompleted
      status: 'True'
      type: ClustersSelected
    - lastTransitionTime: '2022-11-18T16:27:15Z'
      message: Completed validation
      reason: ValidationCompleted
      status: 'True'
      type: Validated
    - lastTransitionTime: '2022-11-18T16:37:16Z'
      message: Policy remediation took too long
      reason: TimedOut
      status: 'False'
      type: Progressing
    - lastTransitionTime: '2022-11-18T16:37:16Z'
      message: Policy remediation took too long
      reason: TimedOut
      status: 'False'
      type: Succeeded <2>
  managedPoliciesForUpgrade:
    - name: talm-policy
      namespace: talm-namespace
  managedPoliciesNs:
    talm-policy: talm-namespace
  remediationPlan:
    - - spoke2
  status:
    currentBatch: 1
    currentBatchRemediationProgress:
      spoke1:
        state: Completed
      spoke2:
        policyIndex: 0

>         state: InProgress

Reviewer: Remove

Author: ok

>     currentBatchStartedAt: '2022-11-18T16:27:16Z'

Reviewer: ```suggestion
    currentBatchStartedAt: '2022-11-18T20:27:16Z'
```

Author: Done

* When the cluster changes to non-compliant with the `inform` policy after being updated
====

[id="selecting_clusters"]
== Selecting clusters

{cgu-operator} builds a remediation plan and selects clusters based on the following fields:

* The `clusterLabelSelector` field specifies the labels of the clusters that you want to update. This consists of a list of the standard label selectors from `k8s.io/apimachinery/pkg/apis/meta/v1`. Each selector in the list uses either label value pairs or label expressions. Matches from each selector are added to the final list of clusters along with the matches from the `clusterSelector` field and the `cluster` field.
* The `clusters` field specifies a list of clusters to update.
* The `canaries` field specifies the clusters for canary updates.
* The `maxConcurrency` field specifies the number of clusters to update in a batch.

You can use the `clusters`, `clusterLabelSelector`, and `clusterSelector` fields together to create a combined list of clusters.

The remediation plan starts with the clusters listed in the `canaries` field. Each canary cluster forms a single-cluster batch.

> .Sample `ClusterGroupUpgrade` CR in the enabled `field` set to `false`

Reviewer: ```suggestion
.Sample `ClusterGroupUpgrade` CR with the enabled `field` set to `false`
```
?

Author: ok

    - talm-policy
  preCaching: false
  remediationStrategy: <4>
    canaries: <5>
        - spoke1
    maxConcurrency: 2 <6>
    timeout: 240
  clusterLabelSelectors: <7>
    - matchExpressions:
      - key: label1
      operator: In
      values:
        - value1a
        - value1b
  batchTimeoutAction: <8>
status: <9>
        computedMaxConcurrency: 2
    conditions:
      - lastTransitionTime: '2022-11-18T16:27:15Z'
        message: All selected clusters are valid
        reason: ClusterSelectionCompleted
        status: 'True'
        type: ClustersSelected <10>
      - lastTransitionTime: '2022-11-18T16:27:15Z'
        message: Completed validation
        reason: ValidationCompleted
        status: 'True'
        type: Validated <11>
      - lastTransitionTime: '2022-11-18T16:37:16Z'
        message: Not enabled
        reason: NotEnabled
        status: 'False'
        type: Progressing
    managedPoliciesForUpgrade:
      - name: talm-policy
        namespace: talm-namespace
    managedPoliciesNs:
      talm-policy: talm-namespace
    remediationPlan:
      - - spoke1
      - - spoke2
        - spoke3
    status:
----
<1> Defines the list of clusters to update.

> <2> The `enable` field is set to `true`.

Reviewer: ```suggestion
<2> The `enable` field is set to `false`.
```

Author: ok

<3> Lists the user-defined set of policies to remediate.
<4> Defines the specifics of the cluster updates.
<5> Defines the clusters for canary updates.
<6> Defines the maximum number of concurrent updates in a batch. The number of remediation batches is the number of canary clusters, plus the number of clusters, except the canary clusters, divided by the maxConcurrency value. The clusters that are already compliant with all the managed policies are excluded from the remediation plan.
<7> Displays the parameters for selecting clusters.
<8> Controls what happens if a batch times out. Possible values are `abort` or `continue`. If unspecified, the default is `continue`.
<9> Displays information about the status of the updates.

> <10> The `ClustersSelected` fields show that all selected clusters are valid.
> <11> The `Validated` fields show that all selected clusters have been validated.

Reviewer: ```suggestion
<10> The `ClustersSelected` condition shows that all selected clusters are valid.
<11> The `Validated` conditions shows that all selected clusters have been validated.
```

Author: changed both to "condition shows"

      - spoke1
      - spoke4
      enable: true
      managedPolicies:
      - policy1-common-cluster-version-policy
      - policy2-common-pao-sub-policy
      remediationStrategy:
        maxConcurrency: 1
        timeout: 240
    status:
      clusters:
        - name: spoke1
          state: complete
        - name: spoke4
          state: complete
      conditions:
      - message: All selected clusters are valid
        reason: ClusterSelectionCompleted
        status: "True"
        type: ClustersSelected
      - message: Completed validation
        reason: ValidationCompleted
        status: "True"
        type: Validated
      - message: All clusters are compliant with all the managed policies
        reason: Completed
        status: "False"
        type: Progressing <1>
      - message: All clusters are compliant with all the managed policies
        reason: Completed
        status: "True"
        type: Succeeded <2>
      managedPoliciesForUpgrade:
      - name: policy1-common-cluster-version-policy
        namespace: default
      - name: policy2-common-pao-sub-policy
        namespace: default
      remediationPlan:
      - - spoke1
      - - spoke4
      status:

>         currentBatchStartedAt: '2022-11-18T16:27:16Z'

Reviewer: ```suggestion
        completedAt: '2022-11-18T16:27:16Z'
```

Author: ok

  enable: true
  managedPolicies:
    - talm-policy
  preCaching: false
  remediationStrategy:
    maxConcurrency: 2
    timeout: 240
status:
  clusters:
    - name: spoke1
      state: complete
    - currentPolicy: <1>
        name: talm-policy
        status: NonCompliant
      name: spoke2
      state: timedout
  computedMaxConcurrency: 2
  conditions:
    - lastTransitionTime: '2022-11-18T16:27:15Z'
      message: All selected clusters are valid
      reason: ClusterSelectionCompleted
      status: 'True'
      type: ClustersSelected
    - lastTransitionTime: '2022-11-18T16:27:15Z'
      message: Completed validation
      reason: ValidationCompleted
      status: 'True'
      type: Validated
    - lastTransitionTime: '2022-11-18T16:37:16Z'
      message: Policy remediation took too long
      reason: TimedOut
      status: 'False'
      type: Progressing
    - lastTransitionTime: '2022-11-18T16:37:16Z'
      message: Policy remediation took too long
      reason: TimedOut
      status: 'False'
      type: Succeeded <2>
  managedPoliciesForUpgrade:
    - name: talm-policy
      namespace: talm-namespace
  managedPoliciesNs:
    talm-policy: talm-namespace
  remediationPlan:

>     - - spoke2

Author: ok

* When the cluster changes to non-compliant with the `inform` policy after being updated
====

> [id="selecting_clusters"]

Reviewer 2: [Every ID in a module](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-in-module-files) needs the `_{context}` variable added, e.g.:

```suggestion
[id="selecting_clusters_{context}"]
```

Author: thanks. added `_{context}` to each ID

You can use the `clusters`, `clusterLabelSelector`, and `clusterSelector` fields together to create a combined list of clusters.

The remediation plan starts with the clusters listed in the `canaries` field. Each canary cluster forms a single-cluster batch.

.Sample `ClusterGroupUpgrade` CR with the enabled `field` set to `false`

[source,yaml]
----
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  creationTimestamp: '2022-11-18T16:27:15Z'
  finalizers:
    - ran.openshift.io/cleanup-finalizer
  generation: 1
  name: talm-cgu
  namespace: talm-namespace
  resourceVersion: '40451823'
  uid: cca245a5-4bca-45fa-89c0-aa6af81a596c
Spec:
  actions:
    afterCompletion:
      deleteObjects: true
    beforeEnable: {}
  backup: false
  clusters: <1>
    - spoke1
  enable: false <2>
  managedPolicies: <3>
    - talm-policy
  preCaching: false
  remediationStrategy: <4>
    canaries: <5>
        - spoke1
    maxConcurrency: 2 <6>
    timeout: 240
  clusterLabelSelectors: <7>
    - matchExpressions:
      - key: label1
      operator: In
      values:
        - value1a
        - value1b
  batchTimeoutAction: <8>
status: <9>
        computedMaxConcurrency: 2

>     conditions:

Author: Issue was with Ln 104 - indentation of `computedMaxConcurrency` was wrong
